### ISSUE TITLE
IRR function

### ISSUE THREAD

**Issue by kman2013** [markrogoyski/math-php#356]
I'm having an issue with the IRR function
I have a payment stream (same payment amount) for 20 years and then it stops - then I have 0 for 30 years showing and a future value at year 50 - this function returns a negative value (-212,27%) but in Excel I get a value of 5.09%. It correctly gave me a value at year 49.
Is there a max number of values it can do?

**Comment by markrogoyski** [comment]
Hi,
Can you provide your input list of values so I can try to reproduce it?
Thanks.

**Comment by kman2013** [comment]
$values = array(-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,250010);

**Comment by markrogoyski** [comment]
Thank you for the example input. I'm able to reproduce it. I'll update this issue when I have an update.

**Comment by kman2013** [comment]
awesome - thank you so much!!

**Comment by kman2013** [comment]
just to add on to the point - I would like to be able to do the values up to 100 years out - so this one stops at year 49 - so the next 51 years of data wouldn't work - if that is possible... :)

**Comment by markrogoyski** [comment]
Hi @kman2013,

Internal rate of return (IRR) is an interesting function.

To get the result you are looking for that matches what Excel comes up with, provide an initial guess different than the default one. For example:

```php
$values = [-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,-1607,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,250010];
$initialGuess = 0.01;

$irr = Finance::irr($values, $initialGuess); // 0.05090132749
```
This will get the result of 0.05090132749 that you got in Excel.

If you don't provide your own initial guess, the function will use a default one. In this case, it used 0.1, and it came up with the answer -2.1227272756716355 (or -212.27%). That seems very different from 0.05090132749 (or 5.09%). The thing about IRR though, they are both correct answers. There is no single "correct" answer in some cases. Excel works the same way and allows you to provide an initial guess to the =IRR function. Excel probably does some other stuff and computes a few of the answers when there are more than one and probably chooses the lowest positive one, maybe.

What you can do in Excel (or MathPHP) is use the =NPV function (`Finance::npv`) to verify that both 5.09% and -212.27% are correct IRR results to your data set. Suppose your data in Excel is in A1 through A51, you can do the following:
`=NPV(0.05090132749, A1:A51)`
`=NPV(-2.12272727567163, A1:A51)`
Both of these will equal 0, which shows that they are both correct answers to what is the IRR of your data set.

This is just how IRR works. The IRR function in R gives multiple answers as a vector when there are multiple answers. Excel and MathPHP just give you one, which seems to be what most people want. Since IRR can have multiple correct answers, I'm not sure how to provide the one the user expects. That is where the user-provided initial guess comes in, which allows you to influence the result you get back. We can investigate what Excel might be doing to pick its answer, but in the meantime, I'd suggest providing an initial guess when using the IRR function.

You may also want to consider using `Finance::mirr` (=MIRR in Excel) instead. MIRR re-invests the return to give the real yield, whereas IRR doesn't reinvest anything. It's not as useful a financial measurement as MIRR.

In summary, there is no bug here. But maybe as an option we could investigate if it is reasonable to mimic what Excel does to pick a single correct answer from multiple answers.

Let me know if you have any questions.

BTW, I've added this case as a unit test in the develop branch: https://github.com/markrogoyski/math-php/blob/develop/tests/Finance/FinanceTest.php#L915
I've also added a unit test with 100 "years" of data and it works fine.

**Comment by kman2013** [comment]
thank you so much for taking the time to look into it
of course it is the guess... I have battled IRR in Excel for years - to the point where I use a code based goal seek solution to arrive at the result instead of using the IRR function at times.
For the initial guess in Excel I've used 0.03 with the most success over the years - I just tried my example(s) again and they work now.
Thank you!!

**Comment by kman2013** [comment]
Hi..
I have a follow up to this question
I have getting a NAN return - in Excel I can calculate the result
values(-14574.6,3015.43) should result in -79.31%
values(-14574.6,-14574.6,9737.3) should result in -54.18%
values(-14574.6,-14574.6,-14574.6,19316.28) should result in -35.61%
this one works:
values(-14574.6,-14574.6,-14574.6,-14574.6,31192.29)  results in -23.5%

for now I am just outputting NA for the result if NAN is calculated. Is it possible to get the result that Excel gets?

**Comment by markrogoyski** [comment]
Hi,
I can take a look.

For reference, can you show what data you have in Excel and how you call the IRR function with that data?

Thanks,
Mark

**Comment by kman2013** [comment]
[irrSample.xlsx](https://github.com/markrogoyski/math-php/files/3842360/irrSample.xlsx)

I made a sample file for you with the first 5 years 
if you need anything else to look at let me know
thanks so much for the fast reply!

**Comment by markrogoyski** [comment]
@kman2013,

This has been addressed and fixed in the latest version, [v1.10.0](https://github.com/markrogoyski/math-php/releases/tag/v1.10.0). You should find that you get the expected IRR values now instead of NANs.

Thanks for bringing this issue to our attention and providing sample data that reproduces the issue.

Mark
